File: apps/write-8-minutes/ios/Anky/Core/WriteBeforeScroll/WritingSessionEngine.swift
import Foundation
#if SWIFT_PACKAGE
import AnkyProtocol
#endif

struct WritingSessionSnapshot: Equatable {
    let protocolText: String
    let reconstructedText: String
    let elapsedMs: Int64
    let lastAcceptedMs: Int64?
    let isStarted: Bool
    let isClosed: Bool

    var hasCompletedSentence: Bool {
        UnlockPolicy.hasCompletedQuickSentence(in: reconstructedText)
    }

    var hasUnlockableWriting: Bool {
        reconstructedText.contains { !$0.isWhitespace && !$0.isNewline }
    }
}

struct WritingSessionEngine {
    private(set) var writer: AnkyWriter
    private(set) var reconstructedText: String

    init() {
        self.writer = AnkyWriter()
        self.reconstructedText = ""
    }

    init(draftText: String) throws {
        let writer = try AnkyWriter(draftText: draftText)
        let parsed = try AnkyParser.parse(draftText)
        self.writer = writer
        self.reconstructedText = AnkyReconstructor.reconstructText(parsed)
    }

    var protocolText: String {
        writer.text
    }

    var elapsedMs: Int64 {
        writer.writingElapsedMs
    }

    var isStarted: Bool {
        writer.isStarted
    }

    var isClosed: Bool {
        writer.isClosed
    }

    var lastAcceptedMs: Int64? {
        writer.lastAcceptedMs
    }

    var hasReachedFullAnky: Bool {
        elapsedMs >= AnkyDuration.completeRitualMs
    }

    var snapshot: WritingSessionSnapshot {
        WritingSessionSnapshot(
            protocolText: protocolText,
            reconstructedText: reconstructedText,
            elapsedMs: elapsedMs,
            lastAcceptedMs: lastAcceptedMs,
            isStarted: isStarted,
            isClosed: isClosed
        )
    }

    @discardableResult
    mutating func accept(_ text: String, at epochMs: Int64) -> [Character] {
        let characters = text.filter(isForwardWritingCharacter)
        guard !characters.isEmpty else {
            return []
        }

        if characters.count == 1 {
            guard let character = characters.first,
                  writer.accept(character, at: epochMs) else {
                return []
            }
            reconstructedText.append(character)
            return [character]
        }

        let eventTimes = syntheticEventTimes(
            characterCount: characters.count,
            insertionEpochMs: epochMs
        )

        var accepted = [Character]()
        for (index, character) in characters.enumerated() {
            guard writer.accept(character, at: eventTimes[index]) else {
                continue
            }
            reconstructedText.append(character)
            accepted.append(character)
        }
        return accepted
    }

    mutating func prepareToResume(at epochMs: Int64) {
        writer.prepareToResume(at: epochMs)
    }

    @discardableResult
    mutating func replaceSuffix(
        keepingPrefixCharacterCount prefixCount: Int,
        with replacementText: String,
        at epochMs: Int64
    ) -> [Character] {
        let accepted = writer.replaceSuffix(
            keepingPrefixCharacterCount: prefixCount,
            with: replacementText,
            at: epochMs
        )
        guard !accepted.isEmpty else {
            return []
        }

        let clampedPrefixCount = min(max(0, prefixCount), reconstructedText.count)
        let prefix = reconstructedText.prefix(clampedPrefixCount)
        reconstructedText = String(prefix) + String(accepted)
        return accepted
    }

    mutating func closeWithTerminalSilence() {
        writer.closeWithTerminalSilence()
    }

    mutating func reset() {
        writer = AnkyWriter()
        reconstructedText = ""
    }

    func silenceElapsedMs(at epochMs: Int64) -> Int64 {
        guard let lastAcceptedMs else {
            return 0
        }
        return max(0, epochMs - lastAcceptedMs)
    }

    private func isForwardWritingCharacter(_ character: Character) -> Bool {
        character != "\n" && character != "\r"
    }

    private func syntheticEventTimes(characterCount: Int, insertionEpochMs: Int64) -> [Int64] {
        guard characterCount > 1,
              let previousAppendMs = writer.lastAcceptedMs else {
            return Array(repeating: insertionEpochMs, count: characterCount)
        }

        let elapsedSinceLastAppend = max(0, insertionEpochMs - previousAppendMs)
        guard elapsedSinceLastAppend > 0 else {
            return Array(repeating: insertionEpochMs, count: characterCount)
        }

        let baseDelta = elapsedSinceLastAppend / Int64(characterCount)
        let remainder = Int(elapsedSinceLastAppend % Int64(characterCount))
        var cursor = previousAppendMs
        return (0..<characterCount).map { index in
            let delta = baseDelta + (index < remainder ? 1 : 0)
            cursor += delta
            return cursor
        }
    }
}

File: apps/android/app/src/main/java/inc/anky/android/core/protocol/WritingSessionEngine.kt
package inc.anky.android.core.protocol

/**
 * Immutable view of the writing session, mirroring iOS WritingSessionSnapshot
 * (ios/Anky/Core/WriteBeforeScroll/WritingSessionEngine.swift).
 */
data class WritingSessionSnapshot(
    val protocolText: String,
    val reconstructedText: String,
    val elapsedMs: Long,
    val lastAcceptedMs: Long?,
    val isStarted: Boolean,
    val isClosed: Boolean,
) {
    /**
     * Mirrors iOS UnlockPolicy.hasCompletedQuickSentence: a completed sentence
     * OR enough words that a sentence is clearly underway. The sentence logic
     * is inlined here until an Android UnlockPolicy exists; keep both in sync.
     */
    val hasCompletedSentence: Boolean
        get() = hasCompletedSentenceIn(reconstructedText) ||
            wordCountIn(reconstructedText) >= QuickPassWordThreshold

    val hasUnlockableWriting: Boolean
        get() = reconstructedText.any { !it.isWhitespace() }

    companion object {
        /** iOS UnlockPolicy.quickPassWordThreshold. */
        const val QuickPassWordThreshold = 6

        fun wordCountIn(text: String): Int =
            text.split(WhitespaceSplitter).count { it.isNotEmpty() }

        /**
         * A completed sentence is a `.`, `!`, or `?` whose nearest preceding
         * non-whitespace character is a letter or number — so `hello`, `...`,
         * `!?`, and `     .` never unlock, while `I am here.` and `Enough!` do.
         */
        fun hasCompletedSentenceIn(text: String): Boolean {
            var lastMeaningfulCharacter: Char? = null
            for (character in text) {
                if (character == '.' || character == '!' || character == '?') {
                    val previous = lastMeaningfulCharacter
                    if (previous != null && (previous.isLetter() || previous.isDigit())) {
                        return true
                    }
                }
                if (!character.isWhitespace()) {
                    lastMeaningfulCharacter = character
                }
            }
            return false
        }

        private val WhitespaceSplitter = Regex("\\s+")
    }
}

/**
 * Wraps AnkyWriter with a live reconstructed-text mirror, ported from iOS
 * WritingSessionEngine (ios/Anky/Core/WriteBeforeScroll/WritingSessionEngine.swift).
 */
class WritingSessionEngine private constructor(
    private var writer: AnkyWriter,
    reconstructedText: String,
) {
    constructor() : this(AnkyWriter(), "")

    var reconstructedText: String = reconstructedText
        private set

    val protocolText: String
        get() = writer.text

    val elapsedMs: Long
        get() = writer.writingElapsedMs

    val isStarted: Boolean
        get() = writer.isStarted

    val isClosed: Boolean
        get() = writer.isClosed

    val lastAcceptedMs: Long?
        get() = writer.lastAcceptedMs

    val hasReachedFullAnky: Boolean
        get() = elapsedMs >= AnkyDuration.CompleteRitualMs

    fun snapshot(): WritingSessionSnapshot =
        WritingSessionSnapshot(
            protocolText = protocolText,
            reconstructedText = reconstructedText,
            elapsedMs = elapsedMs,
            lastAcceptedMs = lastAcceptedMs,
            isStarted = isStarted,
            isClosed = isClosed,
        )

    fun accept(text: String, epochMs: Long): List<String> {
        val glyphs = text
            .filterNot { it == '\n' || it == '\r' }
            .protocolGlyphsOrNull()
            .orEmpty()
        if (glyphs.isEmpty()) return emptyList()

        if (glyphs.size == 1) {
            val glyph = glyphs.first()
            if (!writer.accept(glyph, epochMs)) return emptyList()
            reconstructedText += glyph
            return glyphs
        }

        val eventTimes = syntheticEventTimes(glyphCount = glyphs.size, insertionEpochMs = epochMs)
        val accepted = mutableListOf<String>()
        glyphs.forEachIndexed { index, glyph ->
            if (writer.accept(glyph, eventTimes[index])) {
                reconstructedText += glyph
                accepted += glyph
            }
        }
        return accepted
    }

    fun prepareToResume(epochMs: Long) {
        writer.prepareToResume(epochMs)
    }

    fun replaceSuffix(keepingPrefixGlyphCount: Int, replacementText: String, epochMs: Long): List<String> {
        val accepted = writer.replaceSuffix(
            keepingPrefixGlyphCount = keepingPrefixGlyphCount,
            replacementText = replacementText,
            epochMs = epochMs,
        )
        if (accepted.isEmpty()) return emptyList()

        // The writer rebuilt its protocol lines from the kept prefix plus the
        // replacement, so reconstructing from the protocol text is exactly the
        // iOS prefix + accepted rewrite.
        reconstructedText = runCatching {
            AnkyReconstructor.reconstructText(AnkyParser.parse(writer.text))
        }.getOrDefault(reconstructedText)
        return accepted
    }

    fun closeWithTerminalSilence() {
        writer.closeWithTerminalSilence()
    }

    fun reset() {
        writer = AnkyWriter()
        reconstructedText = ""
    }

    fun silenceElapsedMs(epochMs: Long): Long {
        val lastAcceptedMs = lastAcceptedMs ?: return 0
        return maxOf(0, epochMs - lastAcceptedMs)
    }

    private fun syntheticEventTimes(glyphCount: Int, insertionEpochMs: Long): List<Long> {
        val previousAppendMs = writer.lastAcceptedMs
        if (glyphCount <= 1 || previousAppendMs == null) {
            return List(glyphCount) { insertionEpochMs }
        }

        val elapsedSinceLastAppend = maxOf(0, insertionEpochMs - previousAppendMs)
        if (elapsedSinceLastAppend <= 0) {
            return List(glyphCount) { insertionEpochMs }
        }

        val baseDelta = elapsedSinceLastAppend / glyphCount
        val remainder = (elapsedSinceLastAppend % glyphCount).toInt()
        var cursor = previousAppendMs
        return List(glyphCount) { index ->
            cursor += baseDelta + if (index < remainder) 1 else 0
            cursor
        }
    }

    companion object {
        fun fromDraft(draftText: String): WritingSessionEngine {
            val writer = AnkyWriter.fromDraft(draftText)
            val parsed = AnkyParser.parse(draftText)
            return WritingSessionEngine(
                writer = writer,
                reconstructedText = AnkyReconstructor.reconstructText(parsed),
            )
        }
    }
}
